### PR TITLE
Add match_matrix_tensor op

### PR DESCRIPTION
**Pull request by Aurelius84**
- **match_matrix_tensor**:

Example:
```python
import numpy as np
from paddle.fluid import layers
x_lod_tensor = layers.data(name='x', shape=[10], lod_level=1)
y_lod_tensor = layers.data(name='y', shape=[10], lod_level=1)
out, out_tmp = layers.match_matrix_tensor(x=x_lod_tensor, y=y_lod_tensor, channel_num=3)
```
match_matrix_tensor

### CODE REVIEW

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+void MatchMatrixTensorOP::InferShape(framework::InferShapeContext* ctx) const {
+  PADDLE_ENFORCE(ctx->HasInput("X"),
+                 "X(Input) of MatchMatrix should not be null.");
+  PADDLE_ENFORCE(ctx->HasInput("Y"),
+                 "Y(Input) of MatchMatrix should not be null.");
+  PADDLE_ENFORCE(ctx->HasInput("W"),
+                 "W(Input) of MatchMatrix should not be null.");
+  PADDLE_ENFORCE(ctx->HasOutput("Out"),
+                 "Out(Output) of Fully Connected should not be null.");
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: the error infomation is wrong...

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+    framework::InferShapeContext* ctx) const {
+  PADDLE_ENFORCE(ctx->HasInput("X"),
+                 "Input(X) of SequencePadGradOp should not be null.");
+  PADDLE_ENFORCE(ctx->HasInput("Y"),
+                 "Input(Y) of SequencePadGradOp should not be null.");
+  PADDLE_ENFORCE(ctx->HasInput("W"),
+                 "Input(W) of SequencePadGradOp should not be null.");
+  PADDLE_ENFORCE(ctx->HasInput(framework::GradVarName("Out")),
+                 "Input(Out@GRAD) of SequencePadGradOp should not be null.");
+  //   PADDLE_ENFORCE(ctx->HasInput(framework::GradVarName("Tmp")),
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: remove

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+                   const CBLAS_TRANSPOSE TransA, const CBLAS_TRANSPOSE TransB,
+                   const int M, const int N, const int K, const T alpha,
+                   const T* A, const T* B, const T beta, T* C) {
+  int lda = (TransA == CblasNoTrans) ? K : M;
+  int ldb = (TransB == CblasNoTrans) ? N : K;
+  blas.GEMM(TransA, TransB, M, N, K, alpha, A, lda, B, ldb, beta, C, N);
+}
+
+template <typename DeviceContext, typename T>
+void lego_cpu_gemm_with_lda(const math::BlasT<DeviceContext, T>& blas,
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: change the name

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    auto* bottom_l_data = x->data<T>();
+    auto* bottom_r_data = y->data<T>();
+    auto* t_data = w->data<T>();
+    auto* bottom_l_trans_data = tmp->mutable_data<T>(ctx.GetPlace());
+    memset(bottom_l_trans_data, 0.0,
+           tmp->dims()[0] * tmp->dims()[1] * sizeof(T));
+
+    auto blas = math::GetBlas<platform::CPUDeviceContext, T>(ctx);
+
+    // int M = x->dims()[0], N = dim_t * dim_in, K = dim_in;
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: remove

File: paddle/fluid/operators/match_matrix_tensor_op.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+class MatchMatrixTensorOpGrad : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override;
+
+ protected:
+  //  framework::OpKernelType GetExpectedKernelType(
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.h]: remove

File: paddle/fluid/operators/match_matrix_tensor_op.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+namespace operators {
+using Tensor = framework::Tensor;
+class MatchMatrixTensorOP : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override;
+
+ protected:
+  //  framework::OpKernelType GetExpectedKernelType(
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.h]: remove

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+      the output is a level-1 LodTensor: 
+        level_0: dim_t
+      
+      NOTE: only support 'float32' data type now.
+
+    )DOC");
+}
+
+template <typename DeviceContext, typename T>
+void lego_cpu_gemm(const math::BlasT<DeviceContext, T>& blas,
```
Comment (lanxianghit) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: change the name, maybe you should use the func in 'search_compute.h'

**Review by lanxianghit — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/match_matrix_tensor_op.cc (hunk 3 above)
Comment (Aurelius84) [reply paddle/fluid/operators/match_matrix_tensor_op.cc]: Fixed. I moved the function into search_compute.h header file.

**Review by lanxianghit — DISMISSED** [review]

**Review by lanxianghit — DISMISSED** [review]

**Review by lanxianghit — DISMISSED** [review]

**Review by luotao1 — DISMISSED** [review]
LGTM for `op_use_default_grad_op_maker.spec`

**Review by luotao1 — DISMISSED** [review]

File: paddle/fluid/operators/search_compute.h [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+}
+
+#ifndef TYPE_USE_FLOAT
+#define TYPE_USE_FLOAT
+#endif
+#ifndef USE_SSE
+#define USE_SSE
+#endif
+
+#if defined(TYPE_USE_FLOAT)
```
Comment (luotao1) [inline paddle/fluid/operators/search_compute.h:83]: `TYPE_USE_FLOAT`, `USE_SSE`, 等宏都没有用到。可以确认下这里到底走了哪个分支，删去不用的宏和分支。

File: paddle/fluid/operators/match_matrix_tensor_op.cc [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+namespace paddle {
+namespace operators {
+using Tensor = framework::Tensor;
+using LoDTensor = framework::LoDTensor;
+using LoD = framework::LoD;
+
+void MatchMatrixTensorOP::InferShape(framework::InferShapeContext* ctx) const {
+  PADDLE_ENFORCE(ctx->HasInput("X"),
+                 "X(Input) of MatchMatrix should not be null.");
```
Comment (luotao1) [inline paddle/fluid/operators/match_matrix_tensor_op.cc]: PADDLE_ENFORCE接口需要修改下。

File: paddle/fluid/operators/search_compute.h (hunk 8 above)
Comment (Aurelius84) [reply paddle/fluid/operators/search_compute.h:83]: 删除了不必要的TYPE_USE_DOUBLE分支，去掉了没有用到的函数

File: paddle/fluid/operators/match_matrix_tensor_op.cc (hunk 9 above)
Comment (Aurelius84) [reply paddle/fluid/operators/match_matrix_tensor_op.cc]: 已按照规范修改

**Review by lanxianghit — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

**Review by guoshengCS — APPROVED** [review]

**Pull request merged**